convert2rhel/subscription.py

```diff
@@ -277,28 +287,131 @@ def args(self):
+        loggerinst.debug("Getting a subscription-manager RegisterServer object from dbus")
+        register_server = system_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/RegisterServer")
+        loggerinst.debug("Starting a private DBus to talk to subscription-manager")
+        address = register_server.Start(
+            i18n.SUBSCRIPTION_MANAGER_LOCALE, dbus_interface="com.redhat.RHSM1.RegisterServer"
+        )
+
+        try:
+            # Use the private bus to register the machine
+            loggerinst.debug("Connecting to the private DBus")
+            private_bus = dbus.connection.Connection(address)
+            loggerinst.debug("Getting a subscription-manager Register object from dbus")
+            register_object = private_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/Register")
+
+            if self.password:
+                loggerinst.debug("Registering via username/password: %s" % " ".join(hide_secrets(self.args)))
+                register_object.Register(
+                    self.org or "",
+                    self.username,
+                    self.password,
+                    EMPTY_DBUS_DICT,
+                    self.connection_opts,
+                    i18n.SUBSCRIPTION_MANAGER_LOCALE,
+                    dbus_interface="com.redhat.RHSM1.Register",
+                )
+
+            else:
+                loggerinst.debug("Registering via org/activation_key: %s" % " ".join(hide_secrets(self.args)))
+                register_object.RegisterWithActivationKeys(
+                    self.org,
+                    [self.activation_key],
+                    EMPTY_DBUS_DICT,
+                    self.connection_opts,
+                    i18n.SUBSCRIPTION_MANAGER_LOCALE,
+                    dbus_interface="com.redhat.RHSM1.Register",
+                )
+
+        finally:
+            # Always shut down the private bus
+            loggerinst.debug("Shutting down private DBus instance")
+            register_server.Stop(i18n.SUBSCRIPTION_MANAGER_LOCALE, dbus_interface="com.redhat.RHSM1.RegisterServer")
+
+
+def parse_url(server_url):
+    """Parse a string as a url in a manner mostly compatible with subscription-manager."""
+    # This is an adaptation of what subscription-manager's cli enforces:
+    # https://github.com/candlepin/subscription-manager/blob/main/src/rhsm/utils.py#L112
+
+    # It should look like it has a scheme followed by other information.
+    if not re.match("^[^:]+://.+", server_url):
+        # If there isn't a scheme, add one now
+        server_url = "https://%s" % server_url
+
+    url_parts = urllib.parse.urlsplit(server_url, allow_fragments=False)
+
+    if url_parts.scheme not in ("https", "http"):
+        raise ValueError(
+            "Subscription manager must be accessed over http or https.  %s is not valid" % url_parts.scheme
+        )
 
-        return output, ret_code
+    return url_parts
```

Comment:
Method has no unit tests. If we want to make it private then we are missing a branch for the `ValueError`, can be covered with the exception handling where this method is called

Reply:
I've made this private and added unit_tests.  We're debating some more changes and cornercases in the serverurl so there might be a branch or two left to cover in the unit_tests but at least parse_url() (renamed) and the ValueError catch are being tested now.